library(shiny)
require(shinyjs)
library(shinythemes)
require(shinycssloaders)
library(DESeq2)
library(tidyr)
library(tidyverse)
library(dplyr)
library(DT)
library(PCAtools) #need dev version for ellipses
library('pheatmap')
library('EnhancedVolcano')
library(colourpicker)
library(RColorBrewer)
library(VennDiagram)
library(ggplot2)
library(UpSetR)
library(clusterProfiler)
library(enrichplot)
library(DOSE)
library(msigdbr)
library(stringr)
library(pathview)

tagList(tags$head(tags$style(type = 'text/css','.navbar-brand{display:none;}')),
    fluidPage(theme = shinytheme('superhero'),
            windowTitle = "MaGIC FIRE-Tool",
            titlePanel(
                fluidRow(
                #column(4, img(height =75 , src = "")), if I ever want a logo in there
                column(12, h1(strong('MaGIC FIRE-Tool'), align = 'center')),
                column(12, h2(strong('Fun Interactive RNA-Seq Expression Tool'), align = 'center'))
                ),
                windowTitle = "MaGIC RNA-Seq" ),
                tags$style(type='text/css', '.navbar{font-size:20px;}'),
                tags$style(type='text/css', '.nav-tabs{padding-bottom:20px;}'),
                tags$head(tags$style(".modal-dialog{ width:1300px}")),

        navbarPage(title ="", id='NAVTABS',

        ## Intro Page
##########################################################################################################################################################
            tabPanel('Introduction',
               fluidRow(
                    column(2,
                    ),
                    column(8,
                        column(12, align = "center", 
                            style="margin-bottom:25px;",
                            h2("Introduction")),
                        hr(),
                        column(12, align="center",
                            style="margin-bottom:50px;", 
                            column(2,),
                            column(8,markdown("Welcome to the Fun Interactive RNA-Seq Expression Tool by [the Molecular and Genomics Informatics Core](http://www.bioinformagic.io/).
                                This tool is intended to be used in conjunction with data generated by the Core's pipelines to enable you to dive deeper in to your data!
                                \n All available modules will load once you have completed the prerequisite steps- starting with uploading your data. 
                            ")),
                            column(2,)
                            
                        ),
                        column(12, align='center',
                            icon("upload", "fa-5x", style = "color: #5bc0de")
                        ),
                        column(12, align = "center",
                            style="margin-bottom:25px;", 
                            h3('Step 1: Data Upload')
                        ),
                        fluidRow(
                            style="margin-bottom:75px;",
                            column(6,align = "left",
                                wellPanel(
                                    style="margin-bottom:15px;",
                                    style="height:100%;",
                                    h4('Hit Count File'),
                                    markdown("You should receive this raw hit count table directly from the Core in a directly uploadable format. It should contain headers for each column. 
                                        The first two columns should be the [ensembl stable ID](https://useast.ensembl.org/info/genome/stable_ids/index.html) and [gene symbol](https://www.genenames.org/about/guidelines/) respectively. Subsequent columns are for each individual sample. 
                                        Each row should be a unique gene, followed by each sample's respective raw read counts. From the Core this should be a tab delimited file, but if you open it in other programs that may be modified.
                                        "),
                                    actionButton('hitcount_example','Hit Count Example',class='btn btn-info',style="margin-top:15px;")
                                )
                            ),
                            column(6,align = "left",
                                wellPanel(
                                    style="margin-bottom:15px;",
                                    style="height:100%;",
                                    h4('Metadata File'),
                                    markdown("Your experimental metadata is a table that contains the non-RNAseq data for the experiment.
                                        This can be customized to include a myriad of experimental data for sample grouping. 
                                        Primarily, this will be to define your sample groups and will at minimum contain a column with the sample names (identical to the columns from the gene count table) and respective group. 
                                        As this is extended, metadata correlations can be inferred from the sample variance as well. 
                                        "),
                                    actionButton('metadata_example','Metadata Example',class='btn btn-info',style="margin-top:15px;")
                                )
                            )
                        ),
                        column(12, align='center',
                            icon("calculator", "fa-5x", style = "color: #5bc0de")
                        ),
                        column(12, align = "left", 
                            style="margin-bottom:75px;",
                            h3('Step 2: DESeq2 and assign comparisons', align='center',style="margin-bottom:25px;"),
                            fluidRow(
                                column(6,align = "left",
                                    wellPanel(
                                        style="margin-bottom:15px;",
                                        style="height:100%;",
                                        h4('DESeq Execution'),
                                        markdown("Once you have imported your data, you must then execute the [DESeq2 workflow](https://bioconductor.org/packages/release/bioc/html/DESeq2.html). 
                                            You do this by selecting which metadata column contains your comparators. For example, if you are doing a drug response study- you would have a column defining which drug group each sample is (see the example). You would then select that columns name for DESeq2. 
                                            Once DESeq2 has been performed, you may begin to look at the wholistic data visualization tools- such as basic heatmaps and clustering plots. 
                                            ")
                                    )
                                ),
                                column(6,align = "left",
                                    wellPanel(
                                        style="margin-bottom:15px;",
                                        style="height:100%;",
                                        h4('Comparators'),
                                        markdown("Additionally, you must set what comparisons you wish to view- such as Knock Out vs Wild Type control, or Drug A vs No Drug. This occurs in the comparisons tab of DESeq2. 
                                            Define your comparisons, then allow it to pull out the data as well as assign significance. For each comparison you wish to continue with- click on the button to add comparison to the analysis to allow it to be accesible in other modules.
                                            Once you have defined all of your comparisons, you can then explore full visualization. 
                                            ")
                                    )
                                )
                            )
                        ),
                        column(12, align='center',
                            icon("images", "fa-5x", style = "color: #5bc0de")
                        ),
                        column(12, align='center',
                            h3('Step 3: Visualization',style="margin-bottom:25px;", ),
                            markdown('The visualizations can be then explored by changing the upper tabs. 
                                Each visualization tool has associated customizations available. 
                            '),
                            fluidRow(
                                style="margin-bottom:75px;",
                                style="margin-top:45px;",
                                align='left',
                                fluidRow(
                                    column(6,
                                        wellPanel(
                                            style="margin-bottom:15px;",
                                            style="height:100%;",
                                            h4('Clustering Visualization'),
                                            tabsetPanel(
                                                tabPanel('PCA',
                                                    markdown('PCA is principal component analysis. This is a method of reducing high dimensionality data into 2D or 3D visualizations. The individual principal components capture the variability across samples, with the primary variability captured in PC1 and PC2.
                                                        The closer points are more similar in their gene expression profiles, based on the normalized counts. This is often a good quality control step because if your experiment was properly designed and controlled, the samples should clearly cluster by groups. 
                                                        For more variable samples like human derived, you might expect more variability and not as clear cut a divide. 
                                                        '),
                                                    actionButton('pca_example','PCA Example',class='btn btn-info',style="margin-top:15px;")
                                                ),
                                                tabPanel('Distance Matrices',
                                                    markdown('Distance matrices is another way of measuring sample-to-sample relatedness. Here it will perform euclidean distance between each sample, showing which are more related or more distant.
                                                        Similar to PCA, this should show the anticipated groupings of your samples. 
                                                        '),
                                                    actionButton('dm_example','Distance Matrices Example',class='btn btn-info',style="margin-top:15px;")
                                                ),
                                                tabPanel('Eigen Correlations',
                                                    markdown('The individual vectors that are used to calculate principal components are called Eigen vectors. For eigen correlation analysis, we will be correlating those vectors with various metadata columns.
                                                        For smaller experiments this will most likely be skipped, but for more exploratory studies with expansive metadata this can be very helpful. For example, if you have a large patient study with various clinical features (ie- death during study, weight loss, secondary infections), eigen corerlations will help illuminate what metadata effects were driving the variabilty seen in the data. 
                                                        Do note- this is not meant to replace your sample groupings. You *cannot* replace sample groupings with samples that are more/less similar from output PCA/eigen calculations. This is meant to expand on your sample groupings and provide extra insights. 
                                                        '),
                                                    actionButton('eigen_example','Eigen Correlations Example',class='btn btn-info',style="margin-top:15px;")
                                                )
                                            )
                                        )
                                    ),
                                    column(6,
                                        wellPanel(
                                            style="margin-bottom:15px;",
                                            style="height:100%;",
                                            h4('Comparison Visualizations'),
                                            tabsetPanel(
                                                tabPanel('Volcano Plot',
                                                    markdown('Volcano plots are a staple in RNA-seq analysis. These are meant to visualize the differences seen in your direct comparisons. For example, if you are doing treatment vs control you will be able to see the full spread of each gene- is it higher in treatment, lower in treatment, is it significant etc. 
                                                        This will only show cross-sample comparisons though on a per-comparison basis. 
                                                        '),
                                                    actionButton('volcano_example','Volcano Example',class='btn btn-info',style="margin-top:15px;")
                                                ),
                                                tabPanel('Venn Diagram',
                                                    markdown('Venn diagrams allow effective visualization across comparisons. For example, if you wanted to see what was most similar between DrugA vs Untreated control and DrugB vs Untreated control, you could see what significant genes overlap using venn diagrams.
                                                        This is limited in visualization effectiveness though to 5 or less comparisons
                                                        '),
                                                    actionButton('venn_example','Venn Diagram Example',class='btn btn-info',style="margin-top:15px;")
                                                ),
                                                tabPanel('UpSet Plots',
                                                    markdown('UpSet plots are an alternative to Venn diagrams that allow expansion beyond 5 comparisons. This shows for each combination what is shared across the comparisons.
                                                        '),
                                                    actionButton('upset_example','UpSet Plot Example',class='btn btn-info',style="margin-top:15px;")
                                                )
                                            )
                                        )
                                    )
                                ),
                                fluidRow(
                                    column(6,
                                        wellPanel(
                                            style="margin-bottom:15px;",
                                            style="height:100%;",
                                            h4('Gene Visualizations'),
                                            tabsetPanel(
                                                tabPanel('Heatmaps',
                                                    markdown('Heatmaps are another staple of RNA-seq analysis. These allow you to see individual genes across all samples based on their expression levels. 
                                                        In this case, it can also include sample correlations based on hierarchal clustering. 
                                                        '),
                                                    actionButton('heatmap_example','Heatmap Example',class='btn btn-info',style="margin-top:15px;")
                                                ),
                                                tabPanel('Box Plots',
                                                    markdown('Box plots enable visualization of individual genes across the selected groupings. This shows both the mean of the group, as well as the variance across samples within the group.
                                                        '),
                                                    actionButton('box_example','Box Plot Example',class='btn btn-info',style="margin-top:15px;")
                                                ),
                                                tabPanel('Violin Plots',
                                                    markdown('Violin plots enable visualization of individual genes across the selected groupings. This is very similar to box plots, but for higher sample numbers can more accurately represent the variance by the width of the violin. 
                                                        Note- this will not function for groups with less than 3 samples per group. 
                                                        '),
                                                    actionButton('violin_example','Violin Plot Example',class='btn btn-info',style="margin-top:15px;")
                                                )
                                            )
                                        )
                                    ),
                                    column(6,
                                        wellPanel(
                                            style="margin-bottom:15px;",
                                            style="height:100%;",
                                            h4('Pathway Visualizations'),
                                            tabsetPanel(
                                                tabPanel('Dot Plots',
                                                    markdown('Dot Plots are an excellent visualization tool for GO/KEGG/MSigDB pathway visualization. These show the significant pathway alterations in Enrichment or Over Representation.
                                                        Additionally, this can be viewed in ridge plots or barcharts. This is not currently available for species other than human and mouse. 
                                                        '),
                                                    actionButton('dotplot_example','Dot Plot Example',class='btn btn-info',style="margin-top:15px;")
                                                ),
                                                tabPanel('Enrichment Maps',
                                                    markdown('Enrichment map plots are a useful tool fo visualizing the overlapping gene sets of various networks to more easily view functional outputs. This is not currently available for species other than human and mouse. 
                                                        '),
                                                    actionButton('enrich_example','Enrichment Plot Example',class='btn btn-info',style="margin-top:15px;")
                                                ),
                                                tabPanel('GSEA Plots',
                                                    markdown('GSEA plots are the classic enrichment plots for a single GSEA term. These show the balance of genes across the specific GSEA term. This is not currently available for species other than human and mouse. 
                                                        '),
                                                    actionButton('gsea_example','Violin Plot Example',class='btn btn-info',style="margin-top:15px;")
                                                ),
                                                tabPanel('Pathview Plots',
                                                    markdown('Pathview is an [incredible tool](https://bioconductor.org/packages/release/bioc/html/pathview.html) to visualize your comparison overlaid on a specific KEGG pathway. This is not currently available for species other than human and mouse. 
                                                        '),
                                                    actionButton('pathviewer_example','Violin Plot Example',class='btn btn-info',style="margin-top:15px;")
                                                )
                                            )
                                        )
                                    )
                                )
                            )
                        )
                    ),
                    column(2,
                    )
                ) 
            ),
        ## Input Page
##########################################################################################################################################################
            tabPanel('Input Data',
                fluidRow(
                    column(3,
                        wellPanel(
                            h2("Input Set up", align='center'),
                            markdown('Example data can be found in the [local repository](https://github.com/alemenze/magic-firetool/tree/main/app/docs)'),
                            hr(),
                            h3("Hit Count File"),
                                                       
                            fileInput('datafile', 'Select hit count file',
                                accept=c(
                                    'text/csv',
                                    'text/comma-separated-values, text/plain',
                                    '.csv',
                                    'text/tsv',
                                    'text/tab-separated-values, text/plain',
                                    '.tsv'), multiple=FALSE),
                            radioButtons('separator','Choose Separator of hit count file',
                                c(
                                    Comma=',',
                                    Tab='\t'),
                                selected=','),

                            h3("Metadata file"),
                            fileInput('metadatafile','Select metadata File',
                                accept=c(
                                    'text/csv',
                                    'text/comma-separated-values, text/plain',
                                    '.csv',
                                    'text/tsv',
                                    'text/tab-separated-values, text/plain',
                                    '.tsv'), multiple=FALSE),
                            radioButtons('metaseparator','Choose Separator of metadata file',
                                c(
                                    Comma=',',
                                    Tab='\t'),
                                selected=','),
                            uiOutput('specieschoice')
                            )    
                        ),
                    column(9,
                        tabsetPanel(id='Input Tables',
                            tabPanel(title='Raw Counts', hr(), dataTableOutput('counts_contents')),
                            tabPanel(title='Metadata', hr(), dataTableOutput('meta_contents'))
                        )
                    )
                )
            ),

        ## DESeq Page
##########################################################################################################################################################
            tabPanel('DESeq2',
                fluidRow(
                    column(3,
                        wellPanel(
                            conditionalPanel("input.DESeq=='DESeq Counts'",
                                selectInput("DESeqFactors", label='DESeq Factor for Normalization', choices=NULL),
                                actionButton('DESEQ2',"Execute DESeq2 Workflow",class='btn btn-info')
                            ),
                            conditionalPanel("input.DESeq=='Comparisons'",
                                tags$p("Please select the cross comparisons you wish to view."),
                                tags$p("These are based off of the comparator column chosen for the DESeq normalization."),
                                tags$p("To then add the comparator as an option for later analyses- click the button to 'Add Comparison to Analysis'"),
                                tabPanel(title='OutputComparisons', align='center',
                                    hr(), 
                                    column(12, align='center',actionButton('SaveCompare','Add Current Comparison to Analysis', class='btn btn-info', style="margin-bottom:25px;")),
                                    tags$h3("Current comparisons loaded: "),
                                    textOutput('ListCompares'))
                            )  
                        )
                    ),
                    column(9,
                        tabsetPanel(id='DESeq',
                            tabPanel(title='DESeq Counts', hr(), 
                                withSpinner(type=6, color='#5bc0de',
                                    dataTableOutput('deseq_counts')
                                ),
                                fluidRow(
                                    uiOutput('DESeqDownload') 
                                )
                            ),
                            tabPanel(title='Comparisons',hr(),align='center',
                                fluidRow(style="margin-bottom:15px;",
                                    column(2,),
                                    column(3,selectInput("Comp1", label="Comparison Numerator",choices=NULL)),
                                    column(2, tags$h2("VS")),
                                    column(3, selectInput("Comp2", label="Comparison Denominator",choices=NULL)),
                                    column(2,)
                                ),
                                fluidRow(
                                    column(12, style="margin-bottom:15px;",align='center', actionButton('Compare_groups',"Generate Comparison", class='btn btn-info', style = "width:25%;",style="margin-top:25px;"),
                                        hr())
                                ),
                                column(12, 
                                    withSpinner(type=6, color='#5bc0de',
                                        dataTableOutput('comparison_temp')
                                    ),
                                    
                                ),
                                fluidRow(
                                    uiOutput('ComparisonDownload')
                                )
                            )                          
                        )
                    )
                )
            ),

        ## Clustering Page
##########################################################################################################################################################
            tabPanel('Clustering Plots',
                fluidRow(
                    column(3,
                        wellPanel(
                                conditionalPanel("input.ClusteringPlots=='PCA' || input.ClusteringPlots=='Distance Matrix'",
                                    selectInput("PCAColor", label='Color By', choices=NULL)
                                ),
                                conditionalPanel("input.ClusteringPlots=='PCA'",
                                    selectInput("PCAShape", label='Shape By', choices=NULL),
                                    radioButtons("Ellipsoid",label='Draw Ellipsoids (requires >3 samples/group)',inline=TRUE, choices=c('No'='FALSE','Yes'='TRUE'), selected='FALSE'),
                                    radioButtons("PCALegendPos", label="Legend Position", inline=TRUE, 
                                        choices=c("Top"='top',"Bottom"='bottom','Right'='right','left'='left'), selected="top"),
                                    sliderInput("PCAPointSize","Point Size: ", min=1, max=30, step=1, value=6),
                                    sliderInput("PCALabelSize","Label Size: ", min=1, max=30, step=1, value=5),
                                    sliderInput("PCALegendIcon","Legend Icon Size", min=1, max=30, step=1, value=8),
                                    sliderInput("PCALegendLabel","Legend Label Size", , min=1, max=30, step=1, value=12)
                                ),
                                conditionalPanel("input.ClusteringPlots=='Distance Matrix'",
                                    sliderInput("DMXsize", 'X-Axis Label Size', min=1, max=30, step=1, value=10),
                                    sliderInput("DMYsize", 'Y-Axis Label Size', min=1, max=30, step=1, value=10),
                                    radioButtons("DMang", label='X-Axis Angle', inline=TRUE,
                                        choices=c('0'=0,'45'=45,'90'=90,'270'=270,'315'=315), selected=45)
                                ),
                                conditionalPanel("input.ClusteringPlots=='Eigencorplots'",
                                    selectizeInput("EigenMetavars", label='Select the metadata variables', choices=NULL, multiple=TRUE),
                                    radioButtons("EFun", label='Correlation Function', inline=TRUE,
                                        choices=c("Pearson"='pearson','Spearman'='spearman','Kendall'='kendall'), selected='pearson'),
                                    radioButtons("ERsquare", label="R or R squared", inline=TRUE, 
                                        choices=c("R"='FALSE',"R Squared"='TRUE'), selected="FALSE"),
                                    radioButtons("ECorrect", label="Multiple Testing Correction", inline=TRUE, 
                                        choices=c('False Discovery Rate'='fdr',"Bonferroni"='bonferroni','None'='none'), selected='fdr'),
                                    radioButtons("ELegendPos", label="Legend Position", inline=TRUE, 
                                        choices=c("Top"='top',"Bottom"='bottom','Right'='right','left'='left'), selected="top"),
                                    sliderInput("ELabSize","Value Label size: ", min=1, max=30, step=1, value=2),
                                    sliderInput("ELabColSize","Legend Label Size: ", min=1, max=30, step=1, value=2),
                                    sliderInput("EXLabSize","X-axis Label Size: ", min=1, max=30, step=1, value=2),
                                    sliderInput("EYLabSize","Y-axis Label Size: ", min=1, max=30, step=1, value=2)                                    
                                ),
                                sliderInput('CHeight', label='Plot Heights: ', min=50, max=2000, step=10, value=800),
                                sliderInput('CWidth', label='Plot Widths: ',  min=50, max=2000, step=10, value=800)
                            )    
                        ),
                    column(9,
                        tabsetPanel(id='ClusteringPlots',
                            tabPanel(title='PCA', hr(), 
                                withSpinner(type=6, color='#5bc0de',
                                        plotOutput("pcaplot", height='100%')
                                ),
                                fluidRow(align='center',style="margin-top:25px;",
                                    uiOutput('PCADownload')
                                )
                            ),
                            tabPanel(title='Distance Matrix',hr(), 
                                withSpinner(type=6, color='#5bc0de',
                                        plotOutput("distplot", height='100%')
                                ),
                                fluidRow(align='center',style="margin-top:25px;",
                                    uiOutput('DMDownload')
                                )
                            ),
                            tabPanel(title='Eigencorplots',hr(), 
                                withSpinner(type=6, color='#5bc0de',  
                                    plotOutput('eigencorplotout', height='100%')
                                ),
                                fluidRow(align='center',style="margin-top:25px;",
                                    uiOutput('EigenDownload')
                                )
                            )
        
                        )
                    )
                )
            ),
            
        ## Volcano Page
##########################################################################################################################################################
            tabPanel('Volcano Plots',
                fluidRow(
                    column(3,
                        wellPanel(
                                conditionalPanel("input.VolcanoPlots=='Volcanoes'",
                                    selectInput("VolComp", label="Choose Comparison", choices=NULL),
                                    radioButtons("Vpval", label="FDR adjusted P or p-value", inline=TRUE, 
                                        choices=c("FDR adjusted P"='padj',"P-value"='pvalue'), selected="padj"),
                                    radioButtons("VLabelType", label="Select Label Type", inline=TRUE, 
                                        choices=c("Ensembl Gene"='Gene',"Gene Symbol"='GeneID', 'Selected Genes'='Select','None'='None'), selected="Gene"),
                                    radioButtons("VLegendPos", label="Legend Position", inline=TRUE, 
                                        choices=c("Top"='top',"Bottom"='bottom','Right'='right','left'='left'), selected="top"),
                                    sliderInput("VpointSize","Point size: ", min=1, max=30, step=1, value=5),
                                    sliderInput("VLabSize","Label size: ", min=1, max=30, step=1, value=4),
                                    sliderInput("VLegLabSize","Legend Label size: ", min=1, max=30, step=1, value=10),
                                    sliderInput("VLegIconSize","Legend Icon size: ", min=1, max=30, step=1, value=5),
                                    conditionalPanel("input.VLabelType=='Select'",
                                        selectizeInput("VSelectedGenes", "Please list genes of choice", choices = NULL, multiple=TRUE, options=list(placeholder='Search'))
                                        )                                    
                                ),
                                sliderInput('VHeight', label='Plot Heights: ', min=50, max=2000, step=10, value=800),
                                sliderInput('VWidth', label='Plot Widths: ',  min=50, max=2000, step=10, value=800)
                            )    
                        ),
                    column(9,
                        tabsetPanel(id='VolcanoPlots',
                            tabPanel(title='Volcanoes', hr(), 
                                withSpinner(type=6, color='#5bc0de',  
                                    plotOutput("vol_static", height='100%')
                                ),
                                fluidRow(align='center',style="margin-top:25px;",
                                    uiOutput('VolcanoDownload')
                                )
                            )
                        )
                    )
                )
            ),

        ## Heatmaps
##########################################################################################################################################################
            tabPanel('Heatmap Plots',
                fluidRow(
                    column(3,
                        wellPanel(
                                conditionalPanel("input.HeatPlots=='Heatmap'",
                                    radioButtons("HSubType", label="Select heatmap subsetting", inline=TRUE, 
                                        choices=c("Top N genes"='topn',"Significantly different"='sigdif','Select your own'='Select'), selected="topn"), 
                                    conditionalPanel("input.HSubType=='topn'",
                                        sliderInput("HTopGenes","Number of Top Variable Genes: ", min=1, max=1000, step=1, value=500),
                                        ),
                                    conditionalPanel("input.HSubType=='sigdif'",
                                        radioButtons('HSigP', label='Significance based on: ', inline=TRUE,
                                            choices=c('P Value'='pvalue','FDR P'='padj'), selected='padj')
                                        ),
                                    conditionalPanel("input.HSubType=='Select'",
                                        selectizeInput("HSelectedGenes", "Please list genes of choice", choices = NULL, multiple=TRUE, options=list(placeholder='Search'))
                                        ),
                                    radioButtons('HRowClust', label='Cluster rows: ', inline=TRUE,
                                        choices=c('True'='TRUE','False'='FALSE'), selected='TRUE'),
                                    radioButtons('HColClust', label='Cluster columns: ', inline=TRUE,
                                        choices=c('True'='TRUE','False'='FALSE'), selected='TRUE'),
                                    radioButtons('HScale', label='Scale By: ', inline=TRUE,
                                        choices=c('Row'='row','Column'='column','None'='none'), selected='row'),
                                    radioButtons('HShowNames', label='Show the rownames: ', inline=TRUE,
                                        choices=c('No'='FALSE','Yes'='TRUE'), selected='FALSE'),
                                    sliderInput("HXsize", 'X-Axis Label Size', min=1, max=30, step=1, value=10),
                                    radioButtons("Hang", label='X-Axis Angle', inline=TRUE,
                                        choices=c('0'=0,'45'=45,'90'=90,'270'=270,'315'=315), selected=45)
                                ),
                                sliderInput('HHeight', label='Plot Heights: ', min=50, max=2000, step=10, value=800),
                                sliderInput('HWidth', label='Plot Widths: ',  min=50, max=2000, step=10, value=800)
                            )    
                        ),
                    column(9,
                        tabsetPanel(id='HeatPlots',
                            tabPanel(title='Heatmap', hr(), 
                                withSpinner(type=6, color='#5bc0de',  
                                    plotOutput('heatmap_static', height='100%')
                                ),
                                fluidRow(align='center',style="margin-top:25px;",
                                    uiOutput('HeatmapDownload')
                                )
                            )
                        )
                    )
                )
            ),

        ## Expression Plots
##########################################################################################################################################################
            tabPanel('Gene Expression Plots',
                fluidRow(
                    column(3,
                        wellPanel(
                                selectizeInput("gene_select", label="Name of gene",
                                    choices = NULL, multiple=FALSE, options=list(placeholder='Search')),
                                selectInput("GEGroup", label='Group by', choices=NULL),
                                sliderInput("GELabelSize", 'Label Size', min=1, max=30, step=1, value=5),
                                sliderInput("GETitleSize", 'Title Size', min=1, max=30, step=1, value=10),
                                radioButtons("GEang", label='X-Axis Angle', inline=TRUE,
                                        choices=c('0'=0,'45'=45,'90'=90,'270'=270,'315'=315), selected=45),
                                radioButtons("GELegendPos", label="Legend Position", inline=TRUE, 
                                        choices=c("Top"='top',"Bottom"='bottom','Right'='right','left'='left'), selected="bottom"),
                                sliderInput('GEHeight', label='Plot Heights: ', min=50, max=2000, step=10, value=800),
                                sliderInput('GEWidth', label='Plot Widths: ',  min=50, max=2000, step=10, value=800)
                            )    
                        ),
                    column(9,
                        tabsetPanel(id='GEPlots',
                            tabPanel('Box Plots', hr(), 
                                withSpinner(type=6, color='#5bc0de',  
                                    plotOutput('box_static', height='100%')
                                ),
                                fluidRow(align='center',style="margin-top:25px;",
                                    uiOutput('BoxPlotDownload')
                                )
                            ),
                            tabPanel('Violin Plots', hr(), 
                                withSpinner(type=6, color='#5bc0de',  
                                    plotOutput('violin_static', height='100%')
                                ),
                                fluidRow(align='center',style="margin-top:25px;",
                                    uiOutput('ViolinPlotDownload')
                                )
                            )
                        )
                    )
                )
            ),

        ## Venn Diagrams
##########################################################################################################################################################
            tabPanel('Venn Diagrams',
                fluidRow(
                    column(3,
                        wellPanel(
                                conditionalPanel("input.VennPlots=='Venn Diagrams'",
                                    column(6, selectInput("VennComp1", label="Choose Comparison1", choices=NULL)),
                                    column(6, colourInput("color1", "Pick color1", "#0000FF",palette = "limited")),
                                    column(6, selectInput("VennComp2", label="Choose Comparison2", choices=NULL)),
                                    column(6, colourInput("color2", "Pick color2", "#00FF00",palette = "limited")),
                                    column(6, selectInput("VennComp3", label="Choose Comparison3", choices=NULL)),
                                    column(6, colourInput("color3", "Pick color3", "#FFFF00",palette = "limited")),
                                    column(6, selectInput("VennComp4", label="Choose Comparison4", choices=NULL)),
                                    column(6, colourInput("color4", "Pick color4", "#7FFFD4",palette = "limited")),
                                    column(6, selectInput("VennComp5", label="Choose Comparison5", choices=NULL)),
                                    column(6, colourInput("color5", "Pick color5", "#FF0000",palette = "limited")),
                                    radioButtons("Vennpval", label="FDR adjusted P or p-value", inline=TRUE, 
                                        choices=c("FDR adjusted P"='padj',"P-value"='pvalue'), selected="padj"),
                                    sliderInput("VennFontSize","Font size: ", min=1, max=30, step=1, value=5),
                                    sliderInput("VennLabelFontSize","Label Font size: ", min=1, max=10, step=0.1, value=1),
                                    sliderInput("VennLineThick","Line thickness: ", min=1, max=30, step=1, value=1),
                                    sliderInput("VennLineType","Line Type: ", min=1, max=10, step=1, value=1),
                                    sliderInput("VennAlpha","Color Opacity: ", min=0, max=1, value=0.5)                                    
                                ),
                                conditionalPanel("input.VennPlots=='UpSet Plots'",
                                    selectizeInput("upset_compares", label="Choose comparisons",
                                        choices = NULL, multiple=TRUE, options=list(placeholder='Search')),
                                    radioButtons("Upsetpval", label="FDR adjusted P or p-value", inline=TRUE, 
                                        choices=c("FDR adjusted P"='padj',"P-value"='pvalue'), selected="padj"),
                                    sliderInput("UpSetintsizetitle","Integer title size: ", min=0.1, max=5, step=0.1, value=2),
                                    sliderInput("UpSetintsizeticklabels","Integer tick size: ", min=0.1, max=5, step=0.1, value=2),
                                    sliderInput("UpSetsetsizetitle","Set title size: ", min=0.1, max=5, step=0.1, value=2),
                                    sliderInput("UpSetsetsizeticklabels","Set tick size: ", min=0.1, max=5, step=0.1, value=2),
                                    sliderInput("UpSetsetnames","Names size: ", min=0.1, max=5, step=0.1, value=3),
                                    sliderInput("UpSetnumbersonbar","Numbers on bar size: ", min=0.1, max=5, step=0.1, value=2),
                                    sliderInput("UpSetPointSize","Point size: ", min=0.1, max=10, step=0.1, value=6), 
                                    sliderInput("UpSetLineSize","Line size: ", min=0.1, max=10, step=0.1, value=4)                           
                                ),
                                sliderInput('VennHeight', label='Plot Heights: ', min=50, max=2000, step=10, value=800),
                                sliderInput('VennWidth', label='Plot Widths: ',  min=50, max=2000, step=10, value=800)  
                            )     
                        ),
                    column(9,
                        tabsetPanel(id='VennPlots',
                            tabPanel(title='Venn Diagrams', hr(), 
                                withSpinner(type=6, color='#5bc0de',  
                                    plotOutput("venn_static", height='100%')
                                ),
                                fluidRow(align='center',style="margin-top:25px;",
                                    uiOutput('VennDownload')
                                )
                            ),
                            tabPanel(title='UpSet Plots', hr(), 
                                withSpinner(type=6, color='#5bc0de',  
                                    plotOutput("upset_static", height='100%')
                                ),
                                fluidRow(align='center',style="margin-top:25px;",
                                    uiOutput('UpSetDownload')
                                )
                            )
                        )
                    )
                )
            ),

        ## Pathways Page
##########################################################################################################################################################
            tabPanel('Enrichment Pathways',
                fluidRow(
                    column(3,
                        wellPanel(
                                conditionalPanel("input.Enrichments=='Gene Set Enrichment' || input.Enrichments=='Over Representation'",
                                    selectInput("EnrichComp", label='Choose Enrichment Comparison', choices=NULL)
                                ),

                                conditionalPanel("input.Enrichments=='Gene Set Enrichment'",
                                    h3('GSEA Settings: '),
                                    radioButtons("GSESet", label="Selected gene set: ", inline=TRUE, 
                                        choices=c("Gene ontology"='GO',"KEGG"='kegg','MSigDB'='msigdb'), selected="GO"),
                                    conditionalPanel("input.GSESet=='GO'",
                                        radioButtons("GSEOnt", label='Select Ontology Source: ', inline=TRUE,
                                        choices=c('All'='ALL','Biological Processes'='BP','Molecular Functions'='MF','Cellular Components'='CC'), selected='ALL')
                                    ),
                                    conditionalPanel("input.GSESet=='msigdb'",
                                        radioButtons("GSEMSig", label='Select Ontology Source: ', inline=TRUE,
                                        choices=c('H'='H','C1'='C1','C2'='C2','C3'='C3','C4'='C4','C5'='C5','C6'='C6','C7'='C7'), selected='H')
                                    ), 
                                    sliderInput("NPerm","Number of Permutations: ", min=500, max=10000, step=500, value=1000),      
                                    radioButtons("GSEPVal", label="P value correction: ", inline=TRUE, 
                                        choices=c("None"='none','Bonferroni'='bonferroni','FDR'='fdr'), selected="none"),
                                    h3('Plot Settings: '),
                                    radioButtons("GSEAPlotType", label="Plot Type: ", inline=TRUE, 
                                        choices=c("Dot Plot"='dotplot','Ridge Chart'='ridgechart','Enrichment Map'='emap'), selected="dotplot"), 
                                    sliderInput("NCategories","Number of pathways to show: ", min=5, max=100, step=5, value=10),
                                    sliderInput("GSEFontSize","Font Size: ", min=1, max=30, step=1, value=10),
                                    conditionalPanel("input.GSEAPlotType=='dotplot'",
                                        sliderInput("GSEYWidth","Pathway Label Width: ", min=5, max=100, step=5, value=40)
                                    ),
                                    radioButtons("ShowGSEATable", label="Show GSEA Table: ", inline=TRUE, 
                                        choices=c("No"='no','Yes'='yes'), selected="no"),                  
                                ),

                                conditionalPanel("input.Enrichments=='Over Representation'",
                                    h3('Over Representation Settings: '),
                                    radioButtons("Enrichpval", label="FDR adjusted P or p-value for input selection", inline=TRUE, 
                                        choices=c("FDR adjusted P"='padj',"P-value"='pvalue'), selected="padj"),
                                    radioButtons("ORASet", label="Selected gene set: ", inline=TRUE, 
                                        choices=c("Gene ontology"='GO',"KEGG"='kegg','MSigDB'='msigdb'), selected="GO"),
                                    conditionalPanel("input.ORASet=='GO'",
                                        radioButtons("ORAOnt", label='Select Ontology Source: ', inline=TRUE,
                                        choices=c('All'='ALL','Biological Processes'='BP','Molecular Functions'='MF','Cellular Components'='CC'), selected='ALL')
                                    ),
                                    conditionalPanel("input.ORASet=='msigdb'",
                                        radioButtons("ORAMSig", label='Select Ontology Source: ', inline=TRUE,
                                        choices=c('H'='H','C1'='C1','C2'='C2','C3'='C3','C4'='C4','C5'='C5','C6'='C6','C7'='C7'), selected='H')
                                    ), 
                                    radioButtons("ORAPVal", label="P value correction: ", inline=TRUE, 
                                        choices=c("None"='none','Bonferroni'='bonferroni','FDR'='fdr'), selected="none"),
                                    h3('Plot Settings: '),
                                    radioButtons("ORAPlotType", label="Plot Type: ", inline=TRUE, 
                                        choices=c("Dot Plot"='dotplot','Bar Chart'='barchart','Enrichment Map'='emap'), selected="dotplot"), 
                                    sliderInput("ORANCategories","Number of pathways to show: ", min=5, max=100, step=5, value=10),
                                    sliderInput("ORAFontSize","Font Size: ", min=1, max=30, step=1, value=10),
                                    conditionalPanel("input.ORAPlotType=='dotplot'",
                                        sliderInput("ORAYWidth","Pathway Label Width: ", min=5, max=100, step=5, value=40)
                                    ),
                                    radioButtons("ShowORATable", label="Show Over Representation Table: ", inline=TRUE, 
                                        choices=c("No"='no','Yes'='yes'), selected="no"),                       
                                ),
                                
                                conditionalPanel("input.Enrichments=='Selected Pathways'",
                                    h4('Data will be based on the last run GSE/ORA analysis'),
                                    radioButtons("SelectPlotType", label='Select type of plot', inline=TRUE,
                                        choices=c('GSEA Plot'='gsea','GSEA Kegg Plot'='gsea_kegg','ORA Kegg Plot'='ora_kegg'), selected='gsea'),
                                    conditionalPanel("input.SelectPlotType=='gsea'",
                                        selectizeInput("GSEAPathwayChoice", label='Choose GSEA Pathway of interest', choices=NULL, 
                                            multiple=F, options=list(placeholder='Enter pathway name'))
                                    ),
                                    conditionalPanel("input.SelectPlotType=='gsea_kegg'",
                                        selectizeInput("GSEAKeggChoice", label='Choose GSEA KEGG Pathway of interest', choices=NULL, 
                                            multiple=F, options=list(placeholder='Enter pathway name')),
                                        actionButton("MakePathviewGSEA","Make your Pathview plot", class='btn btn-info')
                                    ),
                                    conditionalPanel("input.SelectPlotType=='ora_kegg'",
                                        selectizeInput("ORAKeggChoice", label='Choose ORA KEGG Pathway of interest', choices=NULL, 
                                            multiple=F, options=list(placeholder='Enter pathway name')),
                                        actionButton("MakePathviewORA","Make your Pathview plot", class='btn btn-info')
                                    )
                                    
                                ),
                                conditionalPanel("input.Enrichments=='Gene Set Enrichment' || input.Enrichments=='Over Representation'",
                                    sliderInput('EnrichHeight', label='Plot Heights: ', min=50, max=2000, step=10, value=800),
                                    sliderInput('EnrichWidth', label='Plot Widths: ',  min=50, max=2000, step=10, value=800)
                                )
                            )    
                        ),
                    column(9,
                        tabsetPanel(id='Enrichments',
                            tabPanel(title='Gene Set Enrichment',hr(),
                                fluidRow(
                                    withSpinner(type=6, color='#5bc0de',  
                                        plotOutput("gseaplot", height='100%')
                                    )
                                ),
                                fluidRow(align='center',style="margin-top:25px;",
                                    uiOutput('GSEAPlotDownload')
                                ),
                                fluidRow(style="margin-top:25px;",
                                    conditionalPanel("input.ShowGSEATable=='yes'",
                                        h3('Gene Set Enrichment Table'),
                                        withSpinner(type=6, color='#5bc0de',  
                                            dataTableOutput("gsea_table")
                                        )
                                    )
                                ),
                                fluidRow(align='center',style="margin-top:25px;",
                                    uiOutput('GSEATableDownload')
                                )
                            ),

                            tabPanel(title='Over Representation',hr(),
                                fluidRow(
                                    withSpinner(type=6, color='#5bc0de',  
                                        plotOutput("oraplot", height='100%')
                                    )
                                ),
                                fluidRow(align='center',style="margin-top:25px;",
                                    uiOutput('ORAPlotDownload')
                                ),
                                fluidRow(style="margin-top:25px;",
                                    conditionalPanel("input.ShowORATable=='yes'",
                                        h3('Over Representation Analysis Table'),
                                        withSpinner(type=6, color='#5bc0de',  
                                            dataTableOutput("ora_table")
                                        )
                                    )
                                ),
                                fluidRow(align='center',style="margin-top:25px;",
                                    uiOutput('ORATableDownload')
                                )
                            ),

                            tabPanel(title='Selected Pathways',hr(),
                                conditionalPanel("input.SelectPlotType=='gsea'",
                                    fluidRow(
                                        withSpinner(type=6, color='#5bc0de',  
                                            plotOutput('gsea_soloplot', height='100%')
                                        )
                                    ),
                                    fluidRow(align='center',style="margin-top:25px;",
                                        uiOutput('GSEASoloDownload')
                                    ),
                                ),
                                conditionalPanel("input.SelectPlotType=='gsea_kegg'",
                                    fluidRow(
                                        withSpinner(type=6, color='#5bc0de',  
                                            plotOutput('gsea_keggplot', height='100%')
                                        )
                                    ),
                                    fluidRow(align='center',style="margin-top:25px;",
                                        uiOutput('GSEAKEGGDownload')
                                    )
                                ),
                                conditionalPanel("input.SelectPlotType=='ora_kegg'",
                                    fluidRow(
                                        withSpinner(type=6, color='#5bc0de',  
                                            plotOutput('ora_keggplot', height='100%')
                                        )
                                    ),
                                    fluidRow(align='center',style="margin-top:25px;",
                                        uiOutput('ORAKEGGDownload')
                                    )
                                )
                            )                             
                        )
                    )
                )
            ),

        ## Footer
##########################################################################################################################################################
            tags$footer(
                wellPanel(
                    fluidRow(
                        column(4, align='center',
                        tags$a(href="https://github.com/alemenze", icon("github", "fa-3x")),
                        tags$h4('GitHub to submit issues/requests')
                        ),
                        column(4, align='center',
                        tags$a(href="http://www.bioinformagic.io/", icon("magic", "fa-3x")),
                        tags$h4('Bioinfor-MaGIC Home Page')
                        ),
                        column(4, align='center',
                        tags$a(href="https://alemenze.github.io/", icon("address-card", "fa-3x")),
                        tags$h4("Developer's Page")
                        )
                    ),
                    fluidRow(
                        column(12, align='center',
                            HTML('<a href="https://www.youtube.com/watch?v=dQw4w9WgXcQ">
                            <p>&copy; 
                                <script language="javascript" type="text/javascript">
                                var today = new Date()
                                var year = today.getFullYear()
                                document.write(year)
                                </script>
                            </p>
                            </a>
                            ')
                        )
                    ) 
                )
            )
        )#Ends navbarPage,
    )#Ends fluidpage
)#Ends tagList
